src/sync.rs

```
@@ -204,9 +221,36 @@ impl SolanaChangeListener {
           // of this method, so we keep looping and listening for interesting
           // notifications.
           if let SolanaMessage::Confirmation { id, result, .. } = message {
-            if let Some(pubkey) = self.pending.get(&id) {
-              self.subscriptions.insert(result, *pubkey); // todo remove from pending
-              debug!("created subscripton {} for {}", &result, &*pubkey);
+            if let Some(sub_request) = self.pending.get(&id) {
+              self.subscriptions.insert(result, *sub_request); // todo remove from pending
+              match *sub_request {
+                SubRequest::Account(account) => {
+                  let (key, acc) =
+                    rpc::get_account(self.client.clone(), account).await?;
+
+                  // only insert when we have no entry, we could have received
+                  // an update
+                  self.accounts.entry(key).or_insert(acc);
+                }
+                SubRequest::Program(program_id) => {
+                  // we have a successful program subscription, so now lets get
+                  // all program accounts and add them to our accounts
+                  let accounts: Vec<_> =
+                    rpc::get_program_accounts(self.client.clone(), &program_id)
+                      .await?;
+
+                  // we only insert the account fetched from RPC
+                  // when no other account exists in the accounts DashMap
+                  // we could have received an update while we where fetching
+                  // the program accounts
+                  for (key, acc) in accounts {
+                    self.accounts.entry(key).or_insert(acc);
+                  }
+                }
+                SubRequest::ReconnectAll => warn!("SubRequest::ReconnectAll"),
```

This probably should not be a warning. maybe `info!`?

---

Maybe we can also silently ignore this message. As its expected, and we are only interested in the Program and Account branches.